pkg/channel/consolidated/reconciler/controller/kafkawatcher.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
+
+	"knative.dev/eventing-kafka/pkg/channel/consolidated/kafka"
+)
+
+var (
+	watchersMtx sync.RWMutex
+	cacheMtx    sync.RWMutex
+	// Hooks into the poll logic for testing
+	after = time.After
+	done  = func() {}
+)
+
+type CGCallBack func()
+type Matcher func(string) bool
+
+type ConsumerGroupWatcher interface {
```

Comment:
Is this the best solution we have? Pinging the kafka admin apis? I wonder if this can cause problems on the long run...

Reply:
Isn't there any mechanism like a long polling event system?

Reply:
I'm afraid I'm unaware of any. what are you thinking of? long polling provided by kafka itself for the consumer groups ?

Reply:
Since this watcher can also be used by KafkaSource and KafkaBroker to potentially poke the same Kafka cluster, I wonder if we should have a dedicated controller periodically updating etcd, or have a proxy serving as a cache layer and supporting push notifications.

Reply:
> if we should have a dedicated controller periodically updating etcd, or have a proxy serving as a cache layer and supporting push notifications.

I guess it exists and it's called strimzi :laughing: We could get some ideas from them regarding this issue, but it sounds like down that path we'll end up representing kafka resources (consumer groups, topics) as CRDs, virtually "cloning" what they're already doing.